extensions/NV/EGL_NV_context_priority_realtime.txt

```diff
@@ -0,0 +1,194 @@
+
+    After the discussion of "Querying EGL_RENDER_BUFFER", add:
+
+    "Querying EGL_CONTEXT_PRIORITY_LEVEL_IMG returns the priority this
+    context was actually created with. Note: this may not be the same
+    as specified at context creation time, due to implementation limits
+    on the number of contexts that can be created at a specific priority
+    level in the system."
+
+Issues
+
+    1)  The EGL_IMG_context_priority extension is written to amend the EGL
+        1.4 specification.  Should this extension amend EGL 1.5 or 1.4?
+
+        RESOLVED:  EGL 1.5 because it is newer and better organized to
+        extend context attributes.
+
+        EGL 1.5 rewrites 3.7.1 "Creating Rendering Contexts" to have subsections
+        for different context attributes.  This extension adds a new such section
+        that includes the EGL_IMG_context_priority attribute and values too.
+
+    2)  Is context priority hint supported for both OpenGL and OpenGL ES contexts?
+
+        RESOLVED:  Yes.
+
+    3)  What is the intended application of the realtime priority level?
+
+        RESOLVED:  One anticipated application is the system compositor
+        for a Head Mounted Display (HMD) requires realtime recomposition
+        for time-warping.
+
+    4)  What action causes a context with realtime priority to preempt
+        other contexts?
+
+        RESOLVED:  Preemption by a context with realtime priority should
+        occur when there are pending rendering commands and an implicit or
+        explicit flush (i.e. glFlush or glFinish) occurs.
+
+    5)  What does "trusted" or "appropriate system privilege level"
+        mean in practice for a Linux-based operating system such as Android?
+
+        RESOLVED: Trusted means an application that has higher privileges
+        than other apps such as having CAP_SYS_NICE capability. On Android
+        such applications have to be registered in advance with the OS;
+        unpriviledged third party app cannot acquire this capability.
+
+        This restriction exists so arbitrary applications do not starve or
+        otherwise compromise the interactivity of the system overall.
+
+    6)  In practice how many realtime priority contexts can exist in a system?
+
+        RESOLVED: Only one realtime priority context should be active at one moment
+        to get best performance.
+
+    7)  Can a context created with a realtime priority hint that is
+        in fact given a realtime priority, subsequently find that realtime
+        priority revoked and, if revoked, can it be restored?
+
+        RESOLVED: No, once a context is created with specific priority level, the
+        priority will not change for lifetime of the context. This means there will
+        not be revoking or restoring of realtime priority to already created context.
+
+    8)  The attrib_list for eglCreateContext could list the
+        EGL_CONTEXT_PRIORITY_REALTIME_NV attribute (or some other
```

Comment:
You mean the `EGL_CONTEXT_PRIORITY_LEVEL_IMG` attribute.

On the issue itself, I feel like giving several priority at once is an API usage error, and should either return an error or be undefined/implementation-defined behaviour.

`or some other attribute` should not be in this extension spec, as it shouldn't change the behaviour of all other specs and extensions

Reply:
>>  I feel like giving several priority at once is an API usage error, and should either return an error or be undefined/implementation-defined behaviour.

Yes and no, it's not a usual behavior to pass same attribute twice with different values but it's not an error too. We've chosen not to return error but use the last value in the array as the one that app intended to be selected.

As far as I know, there is no error returned for other attributes also which are already part of EGL spec, if multiple values are specified for same attribute (multiple attrib-value pairs for same attrib) then there is no language in Spec which says which one should be chosen.

Reply:
@sashinde you still didn't fix the issue - what @1ace was pointing out is that the attribute is EGL_CONTEXT_PRIORITY_LEVEL_IMG which could be listed multiple times (EGL_CONTEXT_PRIORITY_REALTIME_NV could be a value associated with the attribute).

Reply:
@dgkoch  Yes right, I'll just update the patch with below as Issue 8:

    8)  The attrib_list for eglCreateContext could list the
        EGL_CONTEXT_PRIORITY_LEVEL_IMG attribute multiple
        times with different valid values. What happens in this case?

Reply:
that sounds good.  I also agree with @1ace that the behaviour in this case is undefined and what you wrote is just what our implementation happens to do currently.

Reply:
@dgkoch Done, updated Issue and Answer both in latest patch.